keylogger_windows.go

```diff
@@ -0,0 +1,109 @@
+//go:build windows
+// +build windows
+
+package main
+
+import (
+	"syscall"
+	"unicode/utf8"
+	"unsafe"
+
+	"github.com/TheTitanrain/w32"
+)
+
+var (
+	moduser32 = syscall.NewLazyDLL("user32.dll")
+
+	procGetKeyboardLayout     = moduser32.NewProc("GetKeyboardLayout")
+	procGetKeyboardState      = moduser32.NewProc("GetKeyboardState")
+	procToUnicodeEx           = moduser32.NewProc("ToUnicodeEx")
+	procGetKeyboardLayoutList = moduser32.NewProc("GetKeyboardLayoutList")
+	procMapVirtualKeyEx       = moduser32.NewProc("MapVirtualKeyExW")
+	procGetKeyState           = moduser32.NewProc("GetKeyState")
+)
+
+// NewKeyLogger creates a new keylogger depending on
+// the platform we are running on (currently only Windows
+// is supported)
+func NewKeyLogger() KeyLogger {
+	return &KeyLoggerWindows{}
+}
+
+// KeyLoggerWindows represents the keylogger
+type KeyLoggerWindows struct {
+	lastKey int
+}
+
+// GetKey gets the current entered key by the user, if there is any
+func (kl *KeyLoggerWindows) GetKey() Key {
+	activeKey := 0
+	var keyState uint16
+
+	for i := 0; i < 256; i++ {
+		keyState = w32.GetAsyncKeyState(i)
+
+		// Check if the most significant bit is set (key is down)
+		// And check if the key is not a non-char key (except for space, 0x20)
+		if keyState&(1<<15) != 0 && !(i < 0x2F && i != 0x20) && (i < 160 || i > 165) && (i < 91 || i > 93) {
+			activeKey = i
+			break
+		}
+	}
+
+	if activeKey != 0 {
+		if activeKey != kl.lastKey {
+			kl.lastKey = activeKey
+			return kl.ParseKeycode(activeKey, keyState)
+		}
+	} else {
+		kl.lastKey = 0
+	}
+
+	return Key{Empty: true}
+}
+
+// ParseKeycode returns the correct Key struct for a key taking in account the current keyboard settings
+// That struct contains the Rune for the key
+func (kl KeyLoggerWindows) ParseKeycode(keyCode int, keyState uint16) Key {
+	key := Key{Empty: false, Keycode: keyCode}
```

Comment:
I just skimmed through this function -- if we do not really care about what Rune is typed, then we might not even need this part of the code in the future.  
Guess it would be a good point for the app,  _"we do not really spy on what you're typing, we just measure the speed"_

Reply:
Agree